trantor/net/inner/TcpConnectionImpl.cc

```diff
@@ -1461,16 +1497,74 @@ void TcpConnectionImpl::sendFile(FILE *fp, size_t offset, size_t length)
     }
 }
 
+void TcpConnectionImpl::sendStream(
+    std::function<std::size_t(char *, std::size_t)> callback,
+    bool chunked)
+{
+    BufferNodePtr node = std::make_shared<BufferNode>();
+    node->offset_ =
+        0;  // node: range offset, length could also be handled for streams
+    node->fileBytesToSend_ = 1;  // force to > 0 until file sent
+    node->streamCallback_ = callback;
+    node->chunked_ = chunked;
+    if (loop_->isInLoopThread())
+    {
+        std::lock_guard<std::mutex> guard(sendNumMutex_);
+        if (sendNum_ == 0)
+        {
+            writeBufferList_.push_back(node);
+            if (writeBufferList_.size() == 1)
+            {
+                sendFileInLoop(writeBufferList_.front());
+                return;
+            }
+        }
+        else
+        {
+            ++sendNum_;
+            auto thisPtr = shared_from_this();
+            loop_->queueInLoop([thisPtr, node]() {
+                thisPtr->writeBufferList_.push_back(node);
+                {
+                    std::lock_guard<std::mutex> guard1(thisPtr->sendNumMutex_);
+                    --thisPtr->sendNum_;
+                }
+
+                if (thisPtr->writeBufferList_.size() == 1)
+                {
+                    thisPtr->sendFileInLoop(thisPtr->writeBufferList_.front());
+                }
+            });
+        }
+    }
+    else
+    {
+        auto thisPtr = shared_from_this();
+        std::lock_guard<std::mutex> guard(sendNumMutex_);
+        ++sendNum_;
+        loop_->queueInLoop([thisPtr, node]() {
+            LOG_TRACE << "Push sendstream to list";
+            thisPtr->writeBufferList_.push_back(node);
+
+            {
+                std::lock_guard<std::mutex> guard1(thisPtr->sendNumMutex_);
+                --thisPtr->sendNum_;
+            }
+
+            if (thisPtr->writeBufferList_.size() == 1)
+            {
+                thisPtr->sendFileInLoop(thisPtr->writeBufferList_.front());
+            }
+        });
+    }
+}
+
 void TcpConnectionImpl::sendFileInLoop(const BufferNodePtr &filePtr)
 {
     loop_->assertInLoopThread();
-#ifndef _WIN32
-    assert(filePtr->sendFd_ >= 0);
-#else
-    assert(filePtr->sendFp_);
-#endif
+    assert(filePtr->isFile());
 #ifdef __linux__
-    if (!isEncrypted_)
+    if (!isEncrypted_ && !filePtr->chunked_)
```

Comment:
I think here the `callback_` shoud be checked instead of the chunked flag. if the sendStream is called by users with a false trucked flag, the sendfile system call will be called with a invalid fd argument.